internal/policies/apparmor/apparmor_test.go

```diff
@@ -0,0 +1,223 @@
+		"error on read-only .new directory":              {entries: defaultMachineProfile, destAlreadyExists: "machine.new", readOnlyApparmorDir: "machine.new", noParserOutput: true, wantErr: true},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		name := name
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			apparmorDir := t.TempDir()
+			parserCmdOutputFile := filepath.Join(t.TempDir(), "parser-output")
+			apparmorParserCmd := mockApparmorParserCmd(t, parserCmdOutputFile)
+			if tc.noApparmorParser {
+				apparmorParserCmd = []string{"this-definitely-does-not-exist"}
+			}
+			if tc.apparmorParserError {
+				apparmorParserCmd = append(apparmorParserCmd, "-Exit1-")
+			}
+
+			object := "machine"
+			if tc.user {
+				object = "users"
+			}
+
+			if tc.destAlreadyExists != "" {
+				require.NoError(t, os.RemoveAll(apparmorDir), "Setup: can't remove apparmor dir before filing it")
+				require.NoError(t,
+					shutil.CopyTree(
+						filepath.Join("testdata", "ApplyPolicy", "apparmor_dir", object), filepath.Join(apparmorDir, tc.destAlreadyExists),
+						&shutil.CopyTreeOptions{Symlinks: true, CopyFunction: shutil.Copy}),
+					"Setup: can't create initial apparmor dir profiles content")
+				fmt.Println(filepath.Join(apparmorDir, tc.destAlreadyExists))
+			}
+			if tc.readOnlyApparmorDir != "" {
+				testutils.MakeReadOnly(t, filepath.Join(apparmorDir, tc.readOnlyApparmorDir))
+			}
+			sat := sat{err: tc.saveAssetsError}
+
+			m := apparmor.New(apparmorDir, apparmor.WithApparmorParserCmd(apparmorParserCmd))
+			err := m.ApplyPolicy(context.Background(), "ubuntu", !tc.user, tc.entries, sat.mockSaveAssetsTo)
+			if tc.wantErr {
+				// We don't return here as we want to check that the apparmor
+				// dir is in the expected state even in error cases
+				require.Error(t, err, "ApplyPolicy should have failed but didn't")
+			} else {
+				require.NoError(t, err, "ApplyPolicy failed but shouldn't have")
+			}
+
+			// Restore permissions to be able to correctly compare trees
+			if tc.readOnlyApparmorDir != "" {
+				// nolint:gosec //false positive, this is a directory
+				err = os.Chmod(filepath.Join(apparmorDir, tc.readOnlyApparmorDir), 0700)
+				require.NoError(t, err, "can't chmod apparmor dir")
+			}
+			testutils.CompareTreesWithFiltering(t, apparmorDir, filepath.Join("testdata", "ApplyPolicy", "golden", "trees", testutils.NormalizeGoldenName(t, name)), update)
+
+			// Return early if we don't want to check apparmor_parser output for
+			// whatever reason (e.g. command did not execute, returned an error before etc.)
+			if tc.noParserOutput {
+				return
+			}
+
+			// Check that apparmor_parser was called with the expected arguments
+			goldPath := filepath.Join("testdata", "ApplyPolicy", "golden", testutils.NormalizeGoldenName(t, name))
```

Comment:
I would organize things a little bit differently:
- golden/<name of the test>
 and under it:
  - we have the output result from the parser command (with a relevant file name, always the same between tests)
  - then, same kind of fake `etc/apparmor.d` instead of "trees" so that we have an idea where, by default, we project it on the hard drive (if that makes sense), or just `apparmor.d` at least or whatever this is "IRL" :)